File: src/abstract_factory/MazeGame.java
package abstract_factory;

import abstract_factory.base.Door;
import abstract_factory.base.Maze;
import abstract_factory.base.MazeFactory;
import abstract_factory.base.Room;

import static abstract_factory.base.Side.*;

public class MazeGame {
    public static Maze createMaze(MazeFactory factory) {
        Maze maze = factory.makeMaze();
        Room r1 = factory.makeRoom(1);
        Room r2 = factory.makeRoom(2);
        Door door = factory.makeDoor(r1, r2);

        maze.addRoom(r1);
        maze.addRoom(r2);

        r1.setSide(NORTH, factory.makeWall());
        r1.setSide(EAST, door);
        r1.setSide(SOUTH, factory.makeWall());
        r1.setSide(WEST, factory.makeWall());

        r2.setSide(NORTH, factory.makeWall());
        r2.setSide(EAST, factory.makeWall());
        r2.setSide(SOUTH, factory.makeWall());
        r2.setSide(WEST, door);

        return maze;
    }
}

File: src/abstract_factory_kotlin/MazeGame.kt
package abstract_factory_kotlin

import abstract_factory_kotlin.base.Maze
import abstract_factory_kotlin.base.MazeFactory
import abstract_factory_kotlin.base.Side.*

object MazeGame {
    fun createMaze(factory: MazeFactory): Maze {
        val maze = factory.makeMaze()
        val r1 = factory.makeRoom(1)
        val r2 = factory.makeRoom(2)
        val door = factory.makeDoor(r1, r2)

        maze.addRoom(r1)
        maze.addRoom(r2)

        r1.setSide(NORTH, factory.makeWall())
        r1.setSide(EAST, door)
        r1.setSide(SOUTH, factory.makeWall())
        r1.setSide(WEST, factory.makeWall())

        r2.setSide(NORTH, factory.makeWall())
        r2.setSide(EAST, factory.makeWall())
        r2.setSide(SOUTH, factory.makeWall())
        r2.setSide(WEST, door)

        return maze
    }
}
